### PR TITLE
Admin commands for gloss videos

### PR DESCRIPTION
**Pull request by susanodd**
COMMANDS AND FILTERS FOR GLOSS VIDEO ADMIN

**FILTERS**

<details>
- By NME Video

All
True
False

- By Perspective Video

All
True
False

- By Filename Correct

All
True
False

- By Backup Video

All
True
False

</details>

**COMMANDS** (on selection)

<details>
- Rename video files to match type

-- The code checks selected backup and normal video files and renames them accordingly
-- This takes care of old "bak bak" sequences and special 7-char Django characters appearing between the ID and extension.
-- It also makes sure the video extension matches the type of the video file. Previously, backup files with bak bak sequences did not have any video extension.

- Set incorrect NME/Perspective filenames to empty string
-- The code effectively erases the selected objects filename from the video object
-- The empty filename shows up in the admin table
-- Once set to empty, it is possible to delete the incorrect objects without deleting the normal video file
-- The correct normal gloss video has not been altered, because the command merely erased the string
--  other code has been modified to handle the possibility that a videofile object may have no path [`str(videofile)` may be empty]

- Move selected backups to DELETED_FILES_FOLDER location

- Renumber selected backups

</details>

Original setup NOVEMBER 2024 is below. The code has evolved from what it originally was.

<details>

Goal: from a selected set of GlossVideo objects in the admin, delete the video file and the object if the version is a backup.
Reorder the remaining backup videos for the glosses in the set.

@vanlummelhuizen if you're working today on Signbank, the new "video admin" command merely prints what it is doing, it does not actually change anything.
I added a rename to it to get the correct video file extension on the backup file.

Perhaps the command is a bit weird now. Or perhaps it should be split into multiple commands.

The command needs to operate on a selection of GlossVideo objects.
So it does what it's supposed to (although commented out and doesn't), and for the "remaining" objects that were not deleted, it renames them and fixes the version number.

Since the command currently "doesn't do anything other than print stuff" it is harmless to try on the real server.

### CODE REVIEW

**Comment by vanlummelhuizen** [comment]
@susanodd Thank you for this PR and making it not too big. I can see it does not do anything, but I assume it should do something in the future. I guess you first want to see that it is outputting the correct stuff before letting it (re)move and rename files.

Although the PR itself is not big, the functions you wrote are a bit long and sometimes hard to follow. I think there are several places where you could improve this. In function *remove_backups*:

* You can get a distinct set of glosses with `distinct_glosses = Gloss.objects.filter(glossvideo__in=queryset).distinct()` instead of 

```python
    glosses_in_queryset = [obj.gloss for obj in queryset]
    distinct_glosses = list(set(glosses_in_queryset))
```

Also, move that statement to the place where it is used.

* You could  get rid of the `if obj.version == 0:` clause if you do `queryset = queryset.filter(version__gt=0)`.

* This code:
```python
    lookup_backup_files = dict()
    for gloss in distinct_glosses:
        lookup_backup_files[gloss] = GlossVideo.objects.filter(gloss=gloss, version__gt=0).order_by('version', 'id')
    for gloss, videos in lookup_backup_files.items():
```
could be reduced to
```python
    for gloss in distinct_glosses:
        videos = GlossVideo.objects.filter(gloss=gloss, version__gt=0).order_by('version', 'id')
```

* There are ways of reducing the `if DEBUG_VIDEOS:` clauses to one line if there is only a print statement. E.g something like `print('bla') if DEBUG_VIDEOS else None`.

* Because there are two steps, move each step to a separate function.

**Comment by susanodd** [comment]
The improvements are much welcome!

That the code seems in the wrong place has to do with it being evolved several times. I wanted to show the non-backup videos as well in the print statements to look at them as well. The glosses need to be retrieved from the selected GlossVideo objects before anything is deleted.

I will split it into several commands, functions. I'm trying to ascertain how to make it the most useful to clean up the objects.
The admin facilities can either do commands on selected objects or queries / filters. I can try making a filter on the type of video file.

The API does not check the video type. That is done in the browser when you select a file. I wanted to check the webm video type but I can't upload it on Safari. But for some reason, there are webm videos from the API. I had to copy locally a webm file onto an already existing backup file in order to check that it can change the name correctly.

Probably we need some conversion commands, too, in the admin. It is much easier to navigate to files here than on Ubuntu.

**Comment by vanlummelhuizen** [comment]
> The glosses need to be retrieved from the selected GlossVideo objects before anything is deleted.

You are right.

**Comment by susanodd** [comment]
@vanlummelhuizen I added a column to the GlossVideo table in the admin. It shows the "file type" (that returned by "file".)

Lots of the "mp4" files are actually Apple iTunes Video (.M4V) Video. So they should be ".mp4" ?

The conversion "ensure_mp4" was checking for h264 for a very long time and just making them be mp4.

**Comment by susanodd** [comment]
Regarding the requests for code improvement

- There are ways of reducing the if DEBUG_VIDEOS: clauses to one line if there is only a print statement. E.g something like print('bla') if DEBUG_VIDEOS else None.

I prefer not to do this, since it makes it difficult to skip over lines because of lack of indentation.

The other changes I have made.

**Comment by susanodd** [comment]
On signbank-susan, the database is from 20 januari.

@vanlummelhuizen @Woseseltops 

What do you guys think of hiding the "commands" and just merging the "new columns" with master? Those are just read only.
It's possible to filter on wrong filenames.
(It's only one line of code to hide the commands (temporarily until approved).

**Comment by susanodd** [comment]
I added high priority to this issue because there are still lots of NME videos that point to the normal video, even after they were re-uploaded.
 
This issue has a filter on NME video as well as "wrong filename".

I put the most recent database on signbank-susan, where this branch is.
And did the search there. To my surprise, the videos are still messed up. (https://github.com/Signbank/Global-signbank/issues/1453#issuecomment-2631050655)
It's not possible to "witness" whether the videos are fixed or not.

File: signbank/video/admin.py [hunk 1]
```diff
@@ -82,23 +90,311 @@ def matching_file_exists(videofile, key):
...
+            if os.path.exists(video_file_full_path):
+                os.rename(source, destination)
+            if DEBUG_VIDEOS:
+                print('video:admin:rename_extension_videos:videofile.name := ', desired_relative_path)
+            glossvideo.videofile.name = desired_relative_path
+            glossvideo.save()
+
+
+@admin.action(description="Remove selected backups")
+def remove_backups(modeladmin, request, queryset):
```
Comment (Woseseltops) [inline signbank/video/admin.py:314]: Alternative suggestion: instead of deleting, move them to an archive location that we can manually empty later.

File: signbank/video/admin.py [hunk 2]
```diff
@@ -82,23 +90,311 @@ def matching_file_exists(videofile, key):
...
+    import os
+    for obj in queryset.filter(glossvideonme=None, glossvideoperspective=None, version__gt=0):
+        # unlink all the files
+        relative_path = str(obj.videofile)
+        video_file_full_path = os.path.join(WRITABLE_FOLDER, str(obj.videofile))
+        if os.path.exists(video_file_full_path):
+            # remove the video file so the GlossVideo object can be deleted
```
Comment (Woseseltops) [inline signbank/video/admin.py:321]: I'd prefer this as guard clause

File: signbank/video/admin.py [hunk 3]
```diff
@@ -82,23 +90,311 @@ def matching_file_exists(videofile, key):
...
+                print('video:admin:renumber_backups: ', original_version, inx, str(video.videofile))
+            # the version of the gloss video object is updated since objects may have been deleted
+            video.version = inx
+            video.save()
+
+
+@admin.action(description="Set incorrect NME/Perspective filenames to empty string")
+def unlink_files(modeladmin, request, queryset):
+    # allow to erase the filename from an object if it has the wrong format and it is for a subclass video
+    # this is a patch for repairing doubly linked files
```
Comment (Woseseltops) [inline signbank/video/admin.py:364]: Let's make this a doscstring.

File: signbank/video/admin.py [hunk 4]
```diff
@@ -167,6 +463,17 @@ def permissions(self, obj=None):
...
+        # if the file exists, this will display its timestamp in the list view
+        if obj is None or not str(obj.videofile):
+            return ""
+        import os
+        video_file_full_path = os.path.join(WRITABLE_FOLDER, str(obj.videofile))
+        if os.path.exists(video_file_full_path):
+            return video_file_type_extension(video_file_full_path)
+        else:
+            return ""
+
```
Comment (Woseseltops) [inline signbank/video/admin.py:476]: Here too I feel it would make more sense to return `None` or raise an error

**Review by Woseseltops — CHANGES_REQUESTED** [review]
I like this initiative, although it's not 100% clear to me how the commands should be used. Perhaps you can create a page on the wiki for it, along with some examples of what will happen? Let me know if you have something, I will try to organize it for you :)

About code style, I have quite a few smaller stylistic things this time. Don't get discouraged by this! It's all small stuff.

File: signbank/video/convertvideo.py [hunk 5]
```diff
@@ -213,10 +213,30 @@ def make_thumbnail_video(sourcefile, targetfile):
...
+        file_extension = '.mp4'
+    return file_extension
+
+
 def video_file_type_extension(video_file_full_path):
+
+    if not video_file_full_path:
+        return ''
+    if 'glossvideo' not in video_file_full_path:
+        return ''
```
[the remark this answers is not in the document]
Comment (susanodd) [reply signbank/video/convertvideo.py:235]: See above.

We had a problem that some videos ended up in the wrong folder.
The check on 'glossvideo' is to check this.

We want to prevent that a file existence test is checking for the existence of a prefix

(The path is constructed by putting "writable" at the front of the relative path stored in the gloss video object.
So the first is to avoid an existence test on "....writable/" (without any actual video)

I'll document the cases of what it's meant to be testing.

I made it be empty because it can also test whether it's empty by the caller. (More or less the same as an error.)

File: signbank/video/models.py [hunk 6]
```diff
@@ -1036,7 +1051,7 @@ def delete_files(self):
             return
         try:
             os.unlink(file_system_path)
-        except OSError:
+        except (PermissionError, OSError):
```
[the remark this answers is not in the document]
Comment (susanodd) [reply signbank/video/models.py:1054]: That's true. I've been running the code with DEBUG_VIDEOS turned on.

There are actually lots of expressions such as this that involve the file system.

File: signbank/video/admin.py [hunk 7]
```diff
@@ -82,23 +90,311 @@ def matching_file_exists(videofile, key):
...
+        else:
+            return queryset.all()
+
+
+def filename_matches_nme(filename):
+    filename_without_extension, ext = os.path.splitext(os.path.basename(filename))
+    try:
+        if m := re.search(r".+-(\d+)_(nme_\d+|nme_\d+_left|nme_\d+_right)$", filename_without_extension):
+            return True
+        return False
```
[the remark this answers is not in the document]
Comment (susanodd) [reply signbank/video/admin.py:104]: I don't think it returns a Boolean?

File: signbank/video/admin.py [hunk 8]
```diff
@@ -82,23 +90,311 @@ def matching_file_exists(videofile, key):
...
+
+
+class GlossVideoFilenameFilter(admin.SimpleListFilter):
+
+    title = _('Filename Correct')
+    parameter_name = 'filename_correct'
+
+    def lookups(self, request, model_admin):
+        file_exists = tuple((b, b) for b in ('True', 'False'))
```
[the remark this answers is not in the document]
Comment (susanodd) [reply signbank/video/admin.py:145]: Okay, I need to figure out how to do that.

(It's passing the query around that is done by Django Admin. I had trouble with it forgetting the query. I made it explicit everywhere.)

File: signbank/video/admin.py [hunk 9]
```diff
@@ -82,23 +90,311 @@ def matching_file_exists(videofile, key):
...
+    def lookups(self, request, model_admin):
+        file_type = tuple((b, b) for b in ('True', 'False'))
+        return file_type
+
+    def queryset(self, request, queryset):
+
+        def matching_file_type(videofile, key):
+            if not key:
+                return False
+            from pathlib import Path
```
[the remark this answers is not in the document]
Comment (susanodd) [reply signbank/video/admin.py:230]: There is some confusion about the circular imports. PyCharm flags stuff.

File: signbank/video/admin.py (hunk 4 above)
Comment (susanodd) [reply signbank/video/admin.py:476]: I'm pretty sure this was for type-checking. Empty string is also of type string.

By the way, here is an existence test that makes use of str(obj.videofile).
If that is empty, it's merely checking whether WRITABLE_FOLDER exists.

That's why all of those tests are there here, and above.

File: signbank/video/admin.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (vanlummelhuizen) [reply signbank/video/admin.py]: Just FYI: if you want to fully match a string, you could use `re.fullmatch` (see also https://docs.python.org/3/library/re.html#search-vs-match). And if you want True if there is a match and False there is none, cast it to a boolean with `bool()`. Then it becomes:

```python
return bool(re.fullmatch(r".+-(\d+)_(nme_\d+|nme_\d+_left|nme_\d+_right)$", filename_without_extension))
```

Again just FYI, the changes @susanodd made per review are clean enough.

**Comment by susanodd** [comment]
Points

- the code was developed on development servers and locally. No video files. The code was made so that it can run without files.

**Comment by Woseseltops** [comment]
Final request: make the documentation shorter, without losing information

Examples

Before
```python
 # The function is only called if there is no file; the filename does not match any correct pattern, the extension does not match any video file type, and this allows the function to work on the development servers.
```

After
```python
# Only runs if no file exists and the filename is invalid, ensuring compatibility with dev servers.
```

Before
```python
# this is only for documentation purposes in the patterns, it's not a setting
# these were found to work properly on Ubuntu and match older files and older code
```

After
```python
# Documentation-only; works on Ubuntu and matches older files/code.  
```
